[ARMv7] Fix i32 wasm args when useJIT=0

justinmichaud opened the pull request:
#### 21f5de370a78c239379d70b00dc4fc5b471a477e
<pre>
[ARMv7] Fix i32 wasm args when useJIT=0
<a href="https://bugs.webkit.org/show_bug.cgi?id=289752">https://bugs.webkit.org/show_bug.cgi?id=289752</a>

Reviewed by Yusuke Suzuki.

These remaining changes shoud fix the jitless wasm tests on armv7. We just make
the wasm_to_js thunk more carful about the 32-bit JSValue representation.

First, we make the callee load location be an EncodedJSValue. This
shouldn&apos;t change 64-bit platforms, but lets us easily also write the
tag. We also skip some troublesome tests, and fix forceICFailure for
wasm call sites.

* JSTests/microbenchmarks/let-const-tdz-environment-parsing-and-hash-consing-speed.js:
* JSTests/stress/dataview-byte-length-large-array-oob-baseline.js:
* JSTests/stress/dataview-byte-length-large-array-oob.js:
* JSTests/stress/dataview-byte-length-large-array.js:
* JSTests/stress/proxy-revoke.js:
* Source/JavaScriptCore/bytecode/Repatch.cpp:
(JSC::linkMonomorphicCall):
* Source/JavaScriptCore/interpreter/CalleeBits.h:
(JSC::CalleeBits::encodeNullCallee):
* Source/JavaScriptCore/interpreter/StackVisitor.cpp:
(JSC::StackVisitor::readFrame):
* Source/JavaScriptCore/jit/CCallHelpers.h:
(JSC::CCallHelpers::storeWasmCalleeCallee):
* Source/JavaScriptCore/llint/WebAssembly.asm:
* Source/JavaScriptCore/wasm/WasmCallee.h:
* Source/JavaScriptCore/wasm/WasmFormat.cpp:
* Source/JavaScriptCore/wasm/WasmFormat.h:
* Source/JavaScriptCore/wasm/WasmOperations.cpp:
(JSC::Wasm::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/wasm/js/JSWebAssemblyInstance.cpp:
(JSC::JSWebAssemblyInstance::finalizeCreation):
* Source/JavaScriptCore/wasm/js/JSWebAssemblyStruct.h:
* Source/JavaScriptCore/wasm/js/WebAssemblyFunction.cpp:
* Source/JavaScriptCore/wasm/js/WebAssemblyFunction.h:
* Source/JavaScriptCore/wasm/js/WebAssemblyFunctionBase.h:
(JSC::WebAssemblyFunctionBase::boxedWasmCalleeLoadLocation const):
* Source/JavaScriptCore/wasm/js/WebAssemblyModuleRecord.cpp:
(JSC::WebAssemblyModuleRecord::initializeImports):

Canonical link: <a href="https://commits.webkit.org/292645@main">https://commits.webkit.org/292645@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/96051f1069924dfc821159cadf0398acb79bed77

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/96618 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/16232 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/6339 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/101690 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/47138 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/16528 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/24666 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/73635 "Found 74 new test failures: compositing/layer-creation/will-change-layer-creation.html css3/filters/backdrop/backdrop-filter-does-not-size-properly-absolute.html fast/canvas/canvas-bg-multiple-removal.html fast/css/acid2-pixel.html fast/css/acid2.html fast/css/empty-generated-content.html fast/inline/list-marker-float-crash.html fast/inline/list-marker-inside-container-with-margin.html fast/inline/min-content-width-with-hypens.html fast/inline/overflowing-content-with-hypens.html ... (failure)") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/30858 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/99621 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/12430 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/87368 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/53971 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/12189 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/5160 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/46466 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/89291 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/82291 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/5253 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/103714 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/95239 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/23685 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/17265 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/82688 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/23936 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/83416 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/82065 "Found 3 new API test failures: /WebKitGTK/TestPrinting:/webkit/WebKitPrintOperation/print-errors, /WebKitGTK/TestPrinting:/webkit/WebKitPrintOperation/print, /WebKitGTK/TestPrinting:/webkit/WebKitPrintOperation/close-after-print (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20611 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/26714 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/4237 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/17159 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/23648 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/28803 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/118865 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/23307 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/33450 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/26787 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/25048 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/b1e97e851b430ff9a0b04efd4be90f6cf7babbd5. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/42444

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/254cf66e8adc772b246fef8d18541b593b8f06c2. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/42444

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/5df8264f42d8cde37a9882422a0571ea304b1a07. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/42444

justinmichaud commented on Source/JavaScriptCore/wasm/WasmOperations.cpp:
```diff
@@ -303,14 +306,21 @@ JSC_DEFINE_JIT_OPERATION(operationWasmToJSExitMarshalArguments, bool, (void* sp,
     const auto& signature = *typeDefinition.as<FunctionSignature>();
     unsigned argCount = signature.argumentCount();
 
+    // Set up bookkeeping in case we throw an exception here.
```
I am not sure why this was working before; If we threw an exception below, we wouldn't have a valid callee

justinmichaud commented on Source/JavaScriptCore/wasm/js/JSWebAssemblyStruct.h:
```diff
@@ -37,7 +37,7 @@ namespace JSC {
 class JSWebAssemblyInstance;
 
-class JSWebAssemblyStruct final : public WebAssemblyGCObjectBase, private TrailingArray<JSWebAssemblyStruct, uint8_t> {
+class alignas(sizeof(uint64_t)) JSWebAssemblyStruct final : public WebAssemblyGCObjectBase, private TrailingArray<JSWebAssemblyStruct, uint8_t> {
```
Fixes an unaligned load issue on armv7

justinmichaud commented on Source/JavaScriptCore/bytecode/Repatch.cpp:
```diff
@@ -87,6 +87,9 @@ void linkMonomorphicCall(VM& vm, JSCell* owner, CallLinkInfo& callLinkInfo, Code
     CodeBlock* callerCodeBlock = jsDynamicCast<CodeBlock*>(owner); // WebAssembly -> JS stubs don't have a valid CodeBlock.
     ASSERT(owner);
 
+    if (Options::forceICFailure())
```
This was helpful in testing when I added logging to the slow path

Constellation replied:
Put `UNLIKELY` here.

eugeneia commented:
I think the commit message is outdated with regard to the changes?

eugeneia commented on Source/JavaScriptCore/interpreter/StackVisitor.cpp:
```diff
@@ -131,6 +131,16 @@ void StackVisitor::readFrame(CallFrame* callFrame)
+    if (codeBlock->type() < UnlinkedProgramCodeBlockType || codeBlock->type() > CodeBlockType) {
+        dataLogLn("Invalid codeblock type: ", *(JSCell*)codeBlock);
+        dataLogLn("Callee: ", RawPointer(callFrame->unsafeCallee().rawPtr()));
+        // ASSERT_NOT_REACHED();
```
Leftover debugging code?

justinmichaud replied:
Fixed

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/8281e6bdf6d6092e820780145a313b089731e56e. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/42444

justinmichaud commented on Source/JavaScriptCore/llint/WebAssembly.asm:
```diff
@@ -1089,7 +1076,7 @@ if ARM64 or ARM64E
     leap (_llint_default_call_trampoline), t5
 end
-    loadp CallLinkInfo::m_codeBlock[t2], t3
+    move 0, t3
```
We are zeroeing out the codeblock. It should be set by the default call slow path in this case

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/4679dcfcd6c44d6c1c6bc0fb76e05158cac3f7e1. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/42444

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/bd78911174700dba3ba4e399ef7175952fc8a63b. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/42444

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/ee06f30b2b8c2878a3f553eef987942c25c6f1f2. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/42444

Constellation commented on Source/JavaScriptCore/llint/WebAssembly.asm:
```diff
@@ -1089,8 +1093,7 @@ if ARM64 or ARM64E
 else
     leap (_llint_default_call_trampoline), t5
 end
-    loadp CallLinkInfo::m_codeBlock[t2], t3
-    storep t3, (CodeBlock - CallerFrameAndPCSize)[sp]
+    storep 0, (CodeBlock - CallerFrameAndPCSize)[sp]
```
I think this is not correct. We should have a codeBlock from callLinkInfo.

Constellation commented on Source/JavaScriptCore/wasm/WasmOperations.cpp:
```diff
@@ -348,24 +348,31 @@ JSC_DEFINE_JIT_OPERATION(operationWasmToJSExitMarshalArguments, bool, (void* sp,
                     *access.operator()<uint64_t>(calleeFrame, dst) = raw;
             } else {
-                auto raw = *access.operator()<uint64_t>(argumentRegisters, GPRInfo::toArgumentIndex(wasmParam.jsr().payloadGPR()) * sizeof(UCPURegister));
+                auto raw = *access.operator()<uintptr_t>(argumentRegisters, GPRInfo::toArgumentIndex(wasmParam.jsr().payloadGPR()) * sizeof(UCPURegister));
```
We should use UCPURegister

Constellation commented on Source/JavaScriptCore/wasm/WasmOperations.cpp:
```diff
@@ -348,24 +348,31 @@ JSC_DEFINE_JIT_OPERATION(operationWasmToJSExitMarshalArguments, bool, (void* sp,
         case TypeKind::I32: {
             if (wasmParam.isStackArgument()) {
-                uint64_t raw = *access.operator()<uint64_t>(cfr, wasmParam.offsetFromSP() + sizeof(CallerFrameAndPC));
+                uint64_t raw = *access.operator()<uintptr_t>(cfr, wasmParam.offsetFromSP() + sizeof(CallerFrameAndPC));
```
Ditto.

Constellation commented on Source/JavaScriptCore/interpreter/StackVisitor.cpp:
```diff
@@ -131,6 +131,16 @@ void StackVisitor::readFrame(CallFrame* callFrame)
     }
 
+#if ASSERT_ENABLED
+    if (codeBlock->type() < UnlinkedProgramCodeBlockType || codeBlock->type() > CodeBlockType) {
```
Use `!codeBlock->inherits<CodeBlock>()`.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/b35f1a10ba0d0d2e74e7f83508ba1089d3255694. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/42444

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/903b6a2a80fc5c047bf505d7b78f6e31782625b6. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/42444

Constellation commented on Source/JavaScriptCore/wasm/WasmOperations.cpp:
```diff
@@ -418,10 +425,8 @@ JSC_DEFINE_JIT_OPERATION(operationWasmToJSExitMarshalArguments, bool, (void* sp,
-    *access.operator()<uintptr_t>(calleeFrame, CallFrameSlot::callee * sizeof(Register) + TagOffset) = JSValue::CellTag;
-#endif
+    *access.operator()<uint64_t>(calleeFrame, CallFrameSlot::callee * sizeof(Register)) = JSValue::encode(importableFunction->importFunction.get());
+    *access.operator()<uint64_t>(calleeFrame, CallFrameSlot::codeBlock * sizeof(Register)) = 0;
```
Let's remove this since it is not necessary.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/96051f1069924dfc821159cadf0398acb79bed77. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/42444

Constellation commented on Source/JavaScriptCore/wasm/WasmOperations.cpp:
```diff
@@ -455,7 +459,6 @@ JSC_DEFINE_JIT_OPERATION(operationWasmToJSExitMarshalReturnValues, void, (void*
     auto scope = DECLARE_THROW_SCOPE(instance->vm());
 
     constexpr int codeBlockOffset = -0x18;
-
```
Let's remove this change

justinmichaud committed:
[ARMv7] Fix i32 wasm args when useJIT=0
https://bugs.webkit.org/show_bug.cgi?id=289752

Reviewed by Yusuke Suzuki.

These remaining changes shoud fix the jitless wasm tests on armv7. We just make
the wasm_to_js thunk more carful about the 32-bit JSValue representation.

First, we make the callee load location be an EncodedJSValue. This
shouldn't change 64-bit platforms, but lets us easily also write the
tag. We also skip some troublesome tests, and fix forceICFailure for
wasm call sites.

* JSTests/microbenchmarks/let-const-tdz-environment-parsing-and-hash-consing-speed.js:
* JSTests/stress/dataview-byte-length-large-array-oob-baseline.js:
* JSTests/stress/dataview-byte-length-large-array-oob.js:
* JSTests/stress/dataview-byte-length-large-array.js:
* JSTests/stress/proxy-revoke.js:
* Source/JavaScriptCore/bytecode/Repatch.cpp:
(JSC::linkMonomorphicCall):
* Source/JavaScriptCore/interpreter/CalleeBits.h:
(JSC::CalleeBits::encodeNullCallee):
* Source/JavaScriptCore/interpreter/StackVisitor.cpp:
(JSC::StackVisitor::readFrame):
* Source/JavaScriptCore/jit/CCallHelpers.h:
(JSC::CCallHelpers::storeWasmCalleeCallee):
* Source/JavaScriptCore/llint/WebAssembly.asm:
* Source/JavaScriptCore/wasm/WasmCallee.h:
* Source/JavaScriptCore/wasm/WasmFormat.cpp:
* Source/JavaScriptCore/wasm/WasmFormat.h:
* Source/JavaScriptCore/wasm/WasmOperations.cpp:
(JSC::Wasm::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/wasm/js/JSWebAssemblyInstance.cpp:
(JSC::JSWebAssemblyInstance::finalizeCreation):
* Source/JavaScriptCore/wasm/js/JSWebAssemblyStruct.h:
* Source/JavaScriptCore/wasm/js/WebAssemblyFunction.cpp:
* Source/JavaScriptCore/wasm/js/WebAssemblyFunction.h:
* Source/JavaScriptCore/wasm/js/WebAssemblyFunctionBase.h:
(JSC::WebAssemblyFunctionBase::boxedWasmCalleeLoadLocation const):
* Source/JavaScriptCore/wasm/js/WebAssemblyModuleRecord.cpp:
(JSC::WebAssemblyModuleRecord::initializeImports):

Canonical link: https://commits.webkit.org/292645@main

webkit-commit-queue commented:
Committed 292645@main (21f5de370a78): <https://commits.webkit.org/292645@main>

Reviewed commits have been landed. Closing PR #42444 and removing active labels.
